File: src/image_sub_node.cpp
#include <cv_bridge/cv_bridge.h>

#include <opencv2/opencv.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>

class ImageSubscriberNode : public rclcpp::Node {
 public:
  ImageSubscriberNode() : Node("image_subscriber_node") {
    image_subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
        "/eih_camera/image_raw", rclcpp::QoS(100),
        std::bind(&ImageSubscriberNode::image_callback, this,
                  std::placeholders::_1));
  }

 private:
  void image_callback(const sensor_msgs::msg::Image::SharedPtr msg) {
    try {
      cv::Mat image = cv_bridge::toCvCopy(msg, "bgr8")->image;
      cv::Mat resized_image;
      cv::resize(image, resized_image, cv::Size(640, 480));

      cv::imshow("Subscribed Image", resized_image);
      cv::waitKey(1);
    } catch (const cv_bridge::Exception &e) {
      RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
    }
  }

  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_subscription_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ImageSubscriberNode>());
  rclcpp::shutdown();
  return 0;
}

File: tm_eih_cam_client/image_sub_node.py
#!/usr/bin/env python3
"""Image Subscriber Node

Simple node to subscribe and display images from the EIH camera.
"""

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2


class ImageSubscriber(Node):
    """Simple image subscriber for testing"""
    
    def __init__(self):
        super().__init__('image_subscriber')
        
        self.subscription = self.create_subscription(
            Image,
            '/eih_camera/image_raw',
            self.image_callback,
            10
        )
        
        self.bridge = CvBridge()
        self.get_logger().info("Image subscriber initialized")
    
    def image_callback(self, msg):
        """Callback for image messages"""
        try:
            # Convert ROS Image to OpenCV image
            cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding='bgr8')
            
            # Display image
            cv2.imshow('EIH Camera', cv_image)
            cv2.waitKey(1)
            
            self.get_logger().debug(
                f"Received image: {msg.width}x{msg.height}, "
                f"encoding: {msg.encoding}"
            )
        except Exception as e:
            self.get_logger().error(f"Error processing image: {str(e)}")


def main(args=None):
    rclpy.init(args=args)
    node = ImageSubscriber()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        cv2.destroyAllWindows()
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
